Replaced tar to improve consistency of archived bundles

jarnfast opened the pull request:
# What does this change

The tar implementation used previously created tar file headers which included OS and user specific values.
Since they couldn't be overwritten the tar file would be very different when archived on Windows vs Linux - or just by different users in Linux.

This is now replaced by a custom path walk that ensure the tar file headers are created consistently across platforms and users - which is possible due to the simple nature of the porter archives. Compression and actual tar writing is handled by the same modules as before.

Consistency is achieved by hardcoding the uid/gid/mtime/filemode.

I've tested the code on Linux and Windows with bundles ranging in size from 127M to 1.8G. Performance and memory consumption appears to be similar to the previously used tar packager.

Example with archive of whalegap:v0.2.0

```
$ date
Mon 19 Sep 2022 06:32:44 PM UTC

$ porter archive whalegap.tgz --reference ghcr.io/getporter/examples/whalegap:v0.2.0

$ tar tvzf whalegap.tgz
drwxr-xr-x 0/0               0 1970-01-01 00:00 ./
drwxr-xr-x 0/0               0 1970-01-01 00:00 ./artifacts/
drwxr-xr-x 0/0               0 1970-01-01 00:00 ./artifacts/layout/
drwxr-xr-x 0/0               0 1970-01-01 00:00 ./artifacts/layout/blobs/
drwxr-xr-x 0/0               0 1970-01-01 00:00 ./artifacts/layout/blobs/sha256/
-rw-r--r-- 0/0        37711437 1970-01-01 00:00 ./artifacts/layout/blobs/sha256/00bf65475aba8f1077fa9629f088a5f531d645faeccb6acd7a8626c7d896a4c4
-rw-r--r-- 0/0        13792375 1970-01-01 00:00 ./artifacts/layout/blobs/sha256/06da3bad9b830df19c8fef54033fb40422d1aecd478a97127981483209a6382b
-rw-r--r-- 0/0         3950801 1970-01-01 00:00 ./artifacts/layout/blobs/sha256/075c6d318c4bc5afc871c127d7ea9658cb9dfc61129d39d68fe5fd68b211a7f0
... <snipped> ..
-rw-r--r-- 0/0        13943045 1970-01-01 00:00 ./artifacts/layout/blobs/sha256/f202f2b12785fc7f13805cfaba70e9785d8ce9095afc566908900f0c95311108
-rw-r--r-- 0/0             864 1970-01-01 00:00 ./artifacts/layout/index.json
-rw-r--r-- 0/0              37 1970-01-01 00:00 ./artifacts/layout/oci-layout
-rw-r--r-- 0/0            4970 1970-01-01 00:00 ./bundle.json
-rw-r--r-- 0/0             458 1970-01-01 00:00 ./relocation-mapping.json


$ sha256sum whalegap.tgz
27c2cf6ef0dfac6f120290a79def2e3417656d0d44adc80991d864d7c8c398ed *whalegap.tgz

```

# What issue does it fix

Closes #2334

# Notes for the reviewer

I opted for simply adding a tiny and very porter specific file traversal and header generation directly in `archive.go`
Could probably be moved to a util'ish go package somewhere - but I couldn't decide on a good place to put it 😉 

It appears two strategies for debug logging exists (tracing vs. context.Err) - I went with the tracing option.

# Checklist
- [ ] Did you write tests?
- [ ] Did you write documentation?
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://getporter.org/src/CONTRIBUTORS.md

carolynvs commented:
Thanks for the pull request! I wanted to let you know that we will review this after 1.0 has been released. At the moment we aren't merging anything major. I've slotted your change to be one of the first after the 1.0 release.

carolynvs commented on pkg/porter/archive.go:
```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
-	err := filepath.Walk(path,
-		func(subpath string, info os.FileInfo, err error) error {
+func (ex *exporter) createTarHeader(ctx context.Context, path string, file string, fileInfo os.FileInfo) (*tar.Header, error) {
+	log := tracing.LoggerFromContext(ctx)
```
Just a suggestion/fyi but a couple ways to make this easier to trace/troubleshoot would be to either create a span per call to createTarHeader, and then write the relative path and the header name as attributes, or to write a log line per call and then again use attributes on the log.

Right now the log.Debugf calls will print to the console if they are running with verbosity=debug, and it will just be kind of rando info being printed. By using a more structure with a span and attributes, it will be easier for someone to troubleshoot later, without dumping contextless debug lines to the console.

jarnfast replied:
Hmm.. I don't get any debug statement printed when I archive a bundle. They are however added to the OTEL trace (even with verbosity=info) as log statements for the active span:
![image](https://user-images.githubusercontent.com/15995839/211270508-7a72785d-26b3-4d8d-b9d7-11e349608c7a.png)

carolynvs replied:
There is an open issue to fix setting verbosity via a flag, #2505, sorry about that!

carolynvs commented on pkg/porter/archive.go:
```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
+	gzipWriter := gzip.NewWriter(pipeWriter)
+	tarWriter := tar.NewWriter(gzipWriter)
+
+	path := filepath.Clean(filepath.Join(srcPath, "."))
```
Can you add a comment about what this line is doing? The significance of joining "." is a bit lost on me.

jarnfast replied:
It'll remove any trailing `os.PathSeparator` from `srcPath` - I kept it as I don't how if the result from `aferox.TempDir` is consistent across the supported OS'es. It does not appear to be necessary with the "magic" join on windows nor linux.

If I can verify the consistency on a darwin machine I can remove the join - otherwise I'll add a comment stating why it's there.

carolynvs replied:
I ran the tests on my mac and what you have now in the PR works fine

jarnfast replied:
Sorry, didn't write that - I also confirmed the functionality on a mac and thus removed the magic join-with-dot code 😄

carolynvs commented on pkg/porter/archive.go:
```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
+	path := filepath.Clean(filepath.Join(srcPath, "."))
+
+	go func() {
+		ctx, log := tracing.StartSpan(ctx)
```
Since this is in an anonymous function and the span is named after the function, let's switch to using StartSpanWithName

```suggestion
		ctx, log := tracing.StartSpanWithName(ctx, "SPANNAME")
```

carolynvs commented on pkg/porter/archive.go:
```diff
@@ -170,26 +165,131 @@ func (ex *exporter) export() error {
+			defer sourceFile.Close()
+			_, err = io.Copy(tarWriter, sourceFile)
+			if err != nil {
+				return err
```
In general it's helpful to wrap errors from external library calls to provide a bit more context of what we were doing when it failed. For example, here if io.Copy fails, adding some context about the file we were trying to write would make it easier to troubleshoot if it fails.

That way we get an error like "failed to write somefile.json: unexpected EOF" instead of just the EOF without knowing which one failed.

```suggestion
				return fmt.Errorf("failed to copy %s: %w", file, err)
```

carolynvs reviewed (commented):
Sorry for the delay in reviewing this. 1.0 is out the door and I'd love to get this into our next minor release. 👍 

Is there a way that we can test that the new implementation is now consistently producing the same hash? For example, can we update an existing test (or add a new one) that validates that when we archive a test bundle, it matches an expected hash value regardless of the agent OS?

jarnfast commented:
Hi! Now it's my turn to say sorry for the delay ;-)

The comments regarding error handling and traces are all valid points and I'll fix that (hopefully soon).

With regard to the automated testing I did consider added the whalegap bundle's hash (generated on a windows machine) to https://github.com/getporter/porter/blob/main/tests/integration/archive_test.go#L19 - but I wasn't sure the bundle was immutable.

I don't know if running a test (e.g. archive and hash whalegap) on multiple machine architectures and comparing the result would be possible - ok, it's probably possible but I have no idea how to do it with Azure pipelines ;)

carolynvs commented:
How about we add some constants to the test file and check against them based on the current OS?

```
if runtime.GOOS == "windows" {
  // assert the digest matches our expected windows digest
} else if runtime.GOOS == "linux {
  // assert for the expected linux digest
}
```

You can add the one for windows and I can provide the other digests in a follow up PR once yours is merged?

jarnfast commented:
> How about we add some constants to the test file and check against them based on the current OS?
> 
> ```
> if runtime.GOOS == "windows" {
>   // assert the digest matches our expected windows digest
> } else if runtime.GOOS == "linux {
>   // assert for the expected linux digest
> }
> ```
> 
> You can add the one for windows and I can provide the other digests in a follow up PR once yours is merged?

I'm not entirely sure I understand the suggestion.

The general idea is that no matter how many times a bundle is archived or on which OS is done it should produce the same consistent digest.

If the integration tests are run on the all supported GOOS'es the added code in `TestArchive_StableDigest`  will assert exactly that.

On the other hand, if the integration tests only runs on a single GOOS it won't achieve much.
If running the integration tests on all supported GOOS'es is deemed overkill then perhaps a new (slimmer) test suite could be created that will run on all GOOS'es for explicit testing of any cross-OS issues?

carolynvs commented:
Sorry I got myself confused for a sec diving back into this issue after so long. You're correct that the existing test would be sufficient if we were running it against mac/linux/windows.

We have outstanding problems with setting up a CI agent on windows/mac due to licensing restrictions for Docker Desktop and are not currently running our tests on all platforms. I've burnt a significant amount of time getting agents setup and at this point if we want to run multi-platform agents, we need someone else to try to get that working... I'm tapped out on the problem and had to move on.

Let's table it for now and we can always incrementally improve it later.

carolynvs reviewed (approved):
Thank you for implementing this! 💯 And again my apologies that it took so long to be reviewed and merged. I hope you'll contribute again sometime and I promise I'll do better.

jarnfast commented:
> Thank you for implementing this! 💯 And again my apologies that it took so long to be reviewed and merged. I hope you'll contribute again sometime and I promise I'll do better.

And I apologize for not getting back to addressing the review comments earlier - I got sidetracked by other tasks and suddenly autumn had turned to winter ⏳ 😄

carolynvs commented:
I don't have time at the moment to work on Porter but if someone submits a PR, I'm happy to review the proposed change so that bundle.json is easier to extract from the bundle. 👍

jarnfast commented:
What about the `relocation-mapping.json` file, should that also be available in the start of the archive?

I'll see if I can find a few hours next week to look at #2197 - it shouldn't be that much of a hack to ensure `bundle.json` is the first file added to that archive.

Jemgoss commented:
Perhaps the safest thing is simply to place ./artifacts - known to be potentially large - at the end of the archive.

carolynvs commented:
> Perhaps the safest thing is simply to place ./artifacts - known to be potentially large - at the end of the archive.

@Jemgoss That's a great idea. 👍
